push-notifications/ios/Plugin/PushNotificationsPlugin.swift

```diff
@@ -0,0 +1,188 @@
+
+enum PushNotificationError: Error {
+    case tokenParsingFailed
+    case tokenRegistrationFailed
+}
+
+enum LocationPermissions: String {
+    case prompt
+    case denied
+    case granted
+}
+
+@objc(PushNotificationsPlugin)
+public class PushNotificationsPlugin: CAPPlugin {
+    private let notificationDelegateHandler = PushNotificationsDelegate()
+    private var appDelegateRegistrationCalled: Bool = false
+
+    override public func load() {
+        FirebaseApp.configure()
+        self.bridge?.notificationRouter.pushNotificationHandler = self.notificationDelegateHandler
+        self.notificationDelegateHandler.plugin = self
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(self.didRegisterForRemoteNotificationsWithDeviceToken(notification:)),
+                                               name: Notification.Name(Notification.Name.capacitorDidRegisterForRemoteNotifications.self.rawValue),
+                                               object: nil)
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(self.didFailToRegisterForRemoteNotificationsWithError(notification:)),
+                                               name: Notification.Name(Notification.Name.capacitorDidFailToRegisterForRemoteNotifications.self.rawValue),
+                                               object: nil)
+    }
+
+    /**
+     * Register for push notifications
+     */
+    @objc func register(_ call: CAPPluginCall) {
+        DispatchQueue.main.async {
+            UIApplication.shared.registerForRemoteNotifications()
+        }
+        call.resolve()
+    }
+
+    /**
+     * Request notification permission
+     */
+    @objc override public func requestPermissions(_ call: CAPPluginCall) {
+        self.notificationDelegateHandler.requestPermissions { granted, error in
+            guard error == nil else {
+                if let err = error {
+                    call.reject(err.localizedDescription)
+                    return
+                }
+
+                call.reject("unknown error in permissions request")
+                return
+            }
+
+            var result: LocationPermissions = .denied
+
+            if granted {
+                result = .granted
+            }
+
+            call.resolve(["receive": result.rawValue])
+        }
+    }
+
+    /**
+     * Check notification permission
+     */
+    @objc override public func checkPermissions(_ call: CAPPluginCall) {
+        self.notificationDelegateHandler.checkPermissions { status in
+            var result: LocationPermissions = .prompt
+
+            switch status {
+            case .notDetermined:
+                result = .prompt
+            case .denied:
+                result = .denied
+            case .ephemeral, .authorized, .provisional:
+                result = .granted
+            default:
```

Comment:
I've seen `@unknown default:` in other plugins for these which will trigger a compile warning if a new enum is added, I think. Do we want to do that here?